app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java

```diff
@@ -535,4 +465,156 @@ public int setSelectionExpandWholeLines() {
     public boolean indexesValid(int... indexes) {
         return TextViewUtils.inRange(0, length(), indexes);
     }
+
+    static class LineNumbersDrawer {
+
+        private final AppCompatEditText _editor;
+        private final Paint _paint = new Paint();
+
+        private int _defaultPaddingLeft;
+        private static final int LINE_NUMBERS_PADDING_LEFT = 14;
+        private static final int LINE_NUMBERS_PADDING_RIGHT = 10;
+
+        private final Rect _visibleRect = new Rect();
+        private final Rect _lastVisibleRect = new Rect();
+
+        private int _startNumber = 1;
+        private int _maxNumber = 1; // to gauge gutter width
+        private int _maxNumberDigits;
+        private int _lastMaxNumberDigits;
+        private int _lastLayoutLineCount;
+        private float _lastTextSize;
+        private int _numberX;
+        private int _gutterLineX;
+        private final List<Integer> _numberYPositions = new ArrayList<>(); // Y positions of numbers to draw
+
+        // Current visible area
+        private int _top;
+        private int _bottom;
+
+
+        public LineNumbersDrawer(AppCompatEditText editor) {
+            _editor = editor;
+            _editor.getLocalVisibleRect(_lastVisibleRect);
+            _paint.setTextAlign(Paint.Align.RIGHT);
+        }
+
+        public void setDefaultPaddingLeft(int padding) {
+            this._defaultPaddingLeft = padding;
+        }
+
+        public void setTextSize(float textSize) {
+            if (this._lastTextSize != _paint.getTextSize()) {
+                this._lastTextSize = _paint.getTextSize();
+            }
+            _paint.setTextSize(textSize);
+        }
+
+        public int getNumberDigits(int number) {
+            if (_lastMaxNumberDigits != _maxNumberDigits) {
+                _lastMaxNumberDigits = _maxNumberDigits;
+            }
+
+            if (number < 10) {
+                _maxNumberDigits = 1;
+            } else if (number < 100) {
+                _maxNumberDigits = 2;
+            } else if (number < 1000) {
+                _maxNumberDigits = 3;
+            } else if (number < 10000) {
+                _maxNumberDigits = 4;
+            } else {
+                _maxNumberDigits = 5;
+            }
+            return _maxNumberDigits;
+        }
+
+        public boolean isVisibleAreaChanged() {
```

Comment:
See `isScrollSignificant()` https://github.com/gsantner/markor/blob/master/app/src/main/java/net/gsantner/markor/frontend/textview/HighlightingEditor.java#L101-L104

Reply:
I used isOutOfLineNumbersArea() instead of isVisibleAreaChanged() later. About detection of line numbers area, drawing line numbers have some different requirements than highlight. Besides, using `LineNumbersDrawer` class own detection method is also for the sake of improving the cohesion and reduceing coupling.